### DISCUSSION TITLE
intercative visualization from Node.js

### DISCUSSION THREAD

**Discussion by pikaciu22x** [statelyai/xstate#1795] [Q&A] [answered]
Hi,
Is there a way to use visualizer interactively with node.js application? Stringifying machine into JSON loses transition functions and etc. I would love to have some kind of workaround to get the visualizer started and served from node.js with all triggers and some kind of state.

If there is no way to achieve this right now, I would be glad to help to search for a solution

**Comment by davidkpiano** [comment, marked as answer]
Yes, there currently is experimental support: https://github.com/davidkpiano/xstate/pull/1770

Look at https://github.com/davidkpiano/xstate/blob/master/packages/xstate-inspect/examples/server.ts to see how it's implemented in the client side. You can then create a WebSocket receiver on the inspector side:

```js
import { createWebSocketReceiver } from '@xstate/inspect';

const webSocketReceiver = createWebSocketReceiver({
  server: 'localhost:8888',
  protocol: 'ws'
});

webSocketReceiver.subscribe(event => {
  // all receiver events
});
```

See the PR and documentation for more details: https://xstate.js.org/docs/packages/xstate-inspect/#implementing This will soon be updated in the statecharts.io visualizer, but you can also listen to those events and create your own inspector.

**Reply by pikaciu22x** [reply]
Thanks a lot!

**Reply by karneges** [reply]
> Yes, there currently is experimental support: #1770
> 
> Look at https://github.com/davidkpiano/xstate/blob/master/packages/xstate-inspect/examples/server.ts to see how it's implemented in the client side. You can then create a WebSocket receiver on the inspector side:
> 
> ```js
> import { createWebSocketReceiver } from '@xstate/inspect';
> 
> const webSocketReceiver = createWebSocketReceiver({
>   server: 'localhost:8888',
>   protocol: 'ws'
> });
> 
> webSocketReceiver.subscribe(event => {
>   // all receiver events
> });
> ```
> 
> See the PR and documentation for more details: https://xstate.js.org/docs/packages/xstate-inspect/#implementing This will soon be updated in the statecharts.io visualizer, but you can also listen to those events and create your own inspector.

Can yo add small example sever + client?

**Reply by thomasaull** [reply]
@karneges I'm not sure if this is still relevant to you, but after creating the Websocket server (like in the link that was shared earlier) the easiest method for me to get the visualizer is to just open `https://statecharts.io/inspect?server=localhost:8888` (note the `server` GET attribute) in a browser

**Reply by stephenhuh** [reply]
@thomasaull tagging onto this, is there a way for us to use `stately.ai/viz?inspect` instead of statecharts.io?

**Reply by Andarist** [reply]
You should be able to do that since the latest release: https://github.com/statelyai/xstate-viz/releases/tag/v0.7.0

**Reply by stephenhuh** [reply]
just in time @Andarist thank you for your contributions!! (i posted another message here, but i've since resolved it).
